src/subtyping.md

```diff
@@ -1,146 +1,143 @@
-if `'big: 'small` ("big contains small" or "big outlives small"), then `'big` is a subtype
-of `'small`. This is a large source of confusion, because it seems backwards
-to many: the bigger region is a *subtype* of the smaller region. But it makes
-sense if you consider our Animal example: Cat is an Animal *and more*,
-just as `'big` is `'small` *and more*.
+Above, we glossed over the fact that `'static: 'b` implied that `&'static T: &'b T`. This uses a property known as variance.
+It's not always as simple as this example though, to understand that let's try extend this example a bit
 
-Put another way, if someone wants a reference that lives for `'small`,
-usually what they actually mean is that they want a reference that lives
-for *at least* `'small`. They don't actually care if the lifetimes match
-exactly. So it should be ok for us to **forget** that something lives for
-`'big` and only remember that it lives for `'small`.
+```rust,ignore
+fn assign<T>(input: &mut T, val: T) {
+    *input = val;
+}
 
-The meowing dog problem for lifetimes will result in us being able to
-store a short-lived reference in a place that expects a longer-lived one,
-creating a dangling reference and letting us use-after-free.
+fn main() {
+    let mut hello: &'static str = "hello";
+    {
+        let world = String::from("world");
+        assign(&mut hello, &world);
+    }
+    println!("{}", hello); // use after free 😿
+}
+```
 
-It will be useful to note that `'static`, the forever lifetime, is a subtype of
-every lifetime because by definition it outlives everything. We will be using
-this relationship in later examples to keep them as simple as possible.
+In `assign`, we are setting the `hello` reference to point to `world`.
+But then `world` goes out of scope, before the later use of `hello` in the println!
 
-With all that said, we still have no idea how to actually *use* subtyping of lifetimes,
-because nothing ever has type `'a`. Lifetimes only occur as part of some larger type
-like `&'a u32` or `IterMut<'a, u32>`. To apply lifetime subtyping, we need to know
-how to compose subtyping. Once again, we need *variance*.
+This is a classic use-after-free bug!
 
-## Variance
+Our first instinct might be to blame the `assign` impl, but there's really nothing wrong here.
+It shouldn't be surprising that we might want to assign a `T` into a `T`.
 
-Variance is where things get a bit complicated.
+The problem is that we cannot assume that `&mut &'static str` and `&mut &'b str` are compatible.
+This must mean that `&mut &'static str` should **not** be a *subtype* of `&mut &'b str`,
+even if `'static` is a subtype of `'b`.
 
-Variance is a property that *type constructors* have with respect to their
-arguments. A type constructor in Rust is any generic type with unbound arguments.
+Variance is the way that Rust defines the relationships of subtypes through their *type constructor*.
```

Comment:
Could we avoid type constructors and talk of _generic types_ instead? It ought to be clearer. Moreover, it incidentally allows to add the sentence: talking of the variance of a non-generic type makes no sense.

  - We're glossing over `fn(&str) <: fn(&'concrete str)` here, by the way, but I think that's fine